File: prepare_for_interview/question/palindrome.py
# // palindrome
# // 回文の時はTrue,そうでないときはFalse (reverse関数はなし)
# // aba → True
# // abc → False

# // 2.Find palindrome
# //  abcracecarbda →　cec, aceca,racecar



# s = 'test'
# #  print(s == ''.join(reversed(s)))

# print(s == s[::-1])


# def is_palindrome(strings:str)->bool:
#     len_strings = len(strings)
#     if len_strings == 0:
#         return False
#     if len_strings == 1:
#         return True
    
#     start = 0
#     end = len_strings-1
#     while(start<end):
#         if strings[start] != strings[end]:
#             return False
#         start += 1
#         end -=1
#     return True

# if __name__ == '__main__':
#     print(is_palindrome('racecar'))

from typing import Generator


def find_palindrome(strings: str,left:int,right:int)->Generator[str,None,None]:

    while 0<=left and right <= len(strings)-1:
        if strings[left] != strings[right]:
            break

        yield strings[left:right+1]
        left -= 1
        right += 1


def find_all_palindrome(strings: str)->Generator[str,None,None]:

    len_strings = len(strings)

    if len_strings == 0:
        yield
    if len_strings == 1:
        yield strings

    # aba
    # abba
    for i in range(1,len_strings):
        yield from find_palindrome(strings,i-1,i+1)
        yield from find_palindrome(strings,i-1,i)

if __name__ == '__main__':

    for s in find_all_palindrome('gfagafcabachhfrafr'):
        print(s)

File: prepare_for_interview/question/palindrome.cpp
// palindrome
// 回文の時はTrue,そうでないときはFalse (reverse関数は禁止)
// aba → True
// abc → Fals
// 2.Find palindrome
//  abcracecarbda →　cec, aceca,racecar
//  abba

// 2024_05_22もう一度

#include <bits/stdc++.h>
using namespace std;

bool is_palindrome(string &text)
{
    if (text.size() < 2)
        return false;

    string reversed;
    for (auto it = text.rbegin(); it != text.rend(); it++)
    {
        reversed += *it;
    }

    return text == reversed;
}

void push_palindrome(int left_edge, int right_edge, string &text, vector<string> &result)
{
    while (0 <= left_edge && right_edge < text.size() && text[left_edge] == text[right_edge])
    {
        result.push_back(string(text.begin() + left_edge, text.begin() + right_edge + 1));
        left_edge--;
        right_edge++;
    }
}

vector<string> find_palindrome(string &text)
{

    vector<string> result;

    for (int i = 0; i < text.size(); i++)
    {
        // abcba
        int left_edge = i - 1;
        int right_edge = i + 1;

        push_palindrome(left_edge, right_edge, text, result);

        // abba
        left_edge = i - 1;
        right_edge = i;
        push_palindrome(left_edge, right_edge, text, result);
    }

    return result;
}

int main()
{

    string text = "abcbacbbc";
    cout << is_palindrome(text) << endl;

    auto ans = find_palindrome(text);

    for (string a : ans)
    {
        cout << a << ",";
    }

    return 0;
}

// #include <bits/stdc++.h>
// using namespace std;

// bool is_palindrome(string &text)
// {
//     if (text.size() <= 1)
//         return false;

//     string reversed;
//     for (auto it = text.rbegin(); it != text.rend(); it++)
//     {
//         reversed += (*it);
//     }

//     return text == reversed;
// }

// void append_palindrome(vector<string> &result, int left_edge, int right_edge, string &text)
// {
//     while (0 <= left_edge && right_edge < text.size())
//     {
//         if (text[left_edge] == text[right_edge])
//         {
//             result.push_back(string(text.begin() + left_edge, text.begin() + right_edge + 1));
//             left_edge--;
//             right_edge++;
//         }
//         else
//             break;
//     }
// }

// vector<string> find_palindrome(string &text)
// {
//     vector<string> result;
//     for (int i = 0; i < text.size(); i++)
//     {
//         int left_edge = i - 1;
//         int right_edge = i + 1;
//         append_palindrome(result, left_edge, right_edge, text);

//         left_edge = i - 1;
//         right_edge = i;
//         append_palindrome(result, left_edge, right_edge, text);
//     }

//     return result;
// }

// int main()
// {

//     string text = "abcracecarbdaabba";

//     vector<string> ans = find_palindrome(text);

//     for (string str : ans)
//     {
//         cout << str << ",";
//     }

//     return 0;
// }